hadoop-hdds/container-service/src/test/java/org/apache/hadoop/ozone/container/common/impl/TestContainerDataYaml.java

```diff
@@ -179,44 +180,7 @@ public void testIncorrectContainerFile() throws IOException{
       GenericTestUtils.assertExceptionContains("No enum constant", ex);
     }
   }
-
-
```

Comment:
Why this test removed?

Reply:
Very good question, I planned to discuss it.

```
-    // This test is for if we upgrade, and then .container files added by new
-    // server will have new fields added to .container file, after a while we
-    // decided to rollback. Then older ozone can read .container files
-    // created or not.
```

This is a limitation of the current checksum calculation. With this restriction we can never add new fields to the container file. 

I think the proper fix here is removing the test and using the upgrade framework to avoid some situation. This unit test is perfect for the master, temporary, but we should have an option to add new fields if new features are allowed (after finalize).

Today it's not possible as upgrade is not merged, but will be possible with upgrade framework (or just denying all the new EC requests). Therefore, I think it's safe to remove the unit test, move forward, and later add specific upgrade tests for EC.

(cc @avijayanhwx )

Reply:
I understand your point. 
I think we can try the following way to handle this? ( we discussed this point and to keep the discussion open, I am posting the comment here)

How about having additional field which calculates the new checksum including new field. Old checksum field unchanged, so, older clients will go with old checksum and newer clients will check new checksum field? 
Let's see if this can work.

Reply:
Yes, this is possible, but we wouldn't like to introduce new checksum fields when we introduce new fields. It should be done (IMHO) together with another fix which ensures that the new checksum field is always backward compatible (for example because all the fields are used to calculate the checksum).

I totally agree that it should be done, but I think it should be done in a separated issue / patch.

Reply:
I was thinking the newer client will always do checksums for available fields in file, but not from java enum fields. Older client will do as is. 
Also think about the option ignore the additional field when replIndex is 0, which is default. So, for non-existent we can assume that as 0 in newer clients.

Reply:
> Also think about the option ignore the additional field when replIndex is 0, which is default

We can do it, but it doesn't help the compatibility issue, and it requires a bigger refactor on the write code. I checked it, and it requires refactoring `ContainerDataRepresenter` (or the surrounding code) which may require bigger work.

As we need to take care about the compatibility issue related to EC calls anyway, I suggest using this patch as is to move forward, and we will take care about all the compatibility testing after the upgrade framework is merged.

Improving the checksum calculation of container files seems to be a good idea anyway, I created https://issues.apache.org/jira/browse/HDDS-5129 to track the issue (and discuss the best approach as there are multiple approach).

Reply:
>We can do it, but it doesn't help the compatibility issue, and it requires a bigger refactor on the write code. I checked it, and it requires refactoring ContainerDataRepresenter (or the surrounding code) which may require bigger work.

With ignoring replIndex =0, for EC disabled clusters should not have any impact right?

Thanks for creating the separated JIRA for improving.

Reply:
> With ignoring replIndex =0, for EC disabled clusters should not have any impact right?

Sorry, I am not sure if I understood this question.

2. Today we have a generic code to serialize all the white-listed fields. It can be modified to support default values and write out spefic if it has a value different from the default, but it requires some code re-organization.

1. Even if we do this (do not write replIndex to the container file for non EC containers) it doesn't help the backward compatible problem. When a new EC type of container is written (eg. replIndex=1) it couldn't be read by the old cluster code (as it's not known if it's an EC container or not). Therefore, we should enable to write the replIndex only after the finalize step). In this case there is no advantage to write or not write `replIndex` to   the container yaml file for normal containers. (It's more like a code style question, if you think it's worth to add more code re-organization for this, I am fine to add it, but from behavior point of view we will have the same results).

Reply:
I meant that, after EC branch merged, but not using EC containers at all, will have not  any impact at all if we skip writing index. 
Anyway for the EC containers, we need to deal. One question: do we allow old code DN to serve the EC container data?
I am worried that Old DN will not know if there is any EC specific logic added at DN.

Reply:
> Anyway for the EC containers, we need to deal. One question: do we allow old code DN to serve the EC container data?

No. It's not possible as EC data will be written only after finalization which means that old DN code (downgrade) won't be supported any more.

Reply:
Yes. Let's skip the replicationIndex=0 field writing to yaml file in this patch. 
@avijayanhwx If we bump keyvalueContainer version, does the upgrade framework handled to check the version numbers? 

Currently, as you noticed in this JIRA, we need to write one additional field in yaml file for EC. Since it's changing the on disk metadata, we need to worry about compt. While in upgrade in progress, your plan is not to allow new features to be used right? In that case, if you are already having check for keyValue container version, then bumping version would make things cleaner. Could you please comment your thoughts here? Thanks